### PR TITLE
Parser Text API

### PR DESCRIPTION
**Pull request by Josef-Vonasek**
### Pull Request Description

Implements basic API for modification of textual representation

### Important Notes

Although the parser doesn't need to reparse the whole file, it must reparse the top level blocks, since macro system can arbitrarily change nesting of the AST.

### Checklist

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: syntax/graph/src/main/scala/org/enso/syntax/graph/DoubleRepresentation.scala [hunk 1]
```diff
@@ -224,7 +231,49 @@ object ParserUtils {
...
+      if (span < pos.index) Nil else fun(span - line.span, line)
+    }
+    state.setModule(loc, module)
+  }
+
+  def getText(loc: Module.Location): String = state.getModule(loc).show
+
+  def insertText(loc: Module.Location, cursor: TextPosition, text: String) =
+    findAndReplace(loc, cursor) { (pos, line) =>
+      val (line1, line2) = line.show.splitAt(pos + cursor.index)
```
Comment (mwu-tow) [inline syntax/graph/src/main/scala/org/enso/syntax/graph/DoubleRepresentation.scala]: `line1` and `line2` sound like we are dealing with two lines, whereas they are parts of the single line.
I'd consider names like `linePrefix` or sth like that.

**Comment by iamrecursion** [comment]
Can you please label your PRs?

File: lib/flexer/src/main/scala/org/enso/flexer/Parser.scala [hunk 2]
```diff
@@ -163,4 +170,6 @@ object Parser {
     }
   }
 
+  class InternalParserError(offset: Int, message: String) extends Error
```
Comment (mwu-tow) [inline lib/flexer/src/main/scala/org/enso/flexer/Parser.scala]: I'd rather use `Exception` base class (and typename suffix) -- from what I understand Error` is rather meant for unrecoverable issues, while I expect that parser failure is not that serious issue (even if exceptional).

File: syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 3]
```diff
@@ -628,6 +628,18 @@ object AST {
...
+    def findAndReplace(f: Line => Option[List[Line]]): Module = {
+      def go(lines: List[Line]): List[Line] = lines match {
+        case Nil => Nil
+        case l +: ls =>
+          f(l) match {
+            case None        => l +: go(ls)
+            case Some(lines) => lines ++ ls
+          }
+      }
+      Module(List1(go(lines.toList)).get)
```
Comment (mwu-tow) [inline syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Please add documentation comment — it is not clear to me from signature what `f` does. Mostly — that `None` means "retain as-is" and only `Some` does replacing.
And that function may throw if `f` yields no lines.

**Pull request merged**